### PR TITLE
feat(ui): G10.5-T3 Topology dependents/dependencies + path overlays + 30s polling-refresh

### LINKED ISSUE #882: Task: G10.5-T3 Topology dependents/dependencies + path query overlays (subgraph + highlight) + 30s polling-refresh
Parent initiative: #342
Parent goal: #336
Depends-on: #881 (the overlays + polling render onto the Cytoscape graph)
Related: #363 (G9.1 dependents/dependencies/path APIs)

## Target

- Add: `/ui/topology?from=<name>&depth=N` (dependents/dependencies subgraph) + `/ui/topology?from=A&to=B` (path) query handling + HTMX `hx-trigger="every 30s"` polling-refresh on the graph container.
- Add: `backend/tests/test_ui_topology_queries.py`.
- Precedent: #881 graph view; the G9.1 `dependents`/`dependencies`/`path` query verbs (`api/v1/topology.py` + `cli/internal/cmd/topology/dependencies.go`/`path.go`, [#363](https://github.com/evoila/meho/issues/363)).

## Context

Per #342 work-items #4, #5, #8 — the query overlays on the graph. Dependents/dependencies (`?from=<name>&depth=N`) wrap the G9.1 APIs; the server returns the subgraph JSON, Cytoscape re-renders. Path (`?from=A&to=B`) wraps `meho topology path`; the server returns the path, Cytoscape highlights it (`cy.$('#A, #B, ...edges').addClass('highlight')`). v0.2 ships polling-refresh (`hx-trigger="every 30s"`); live SSE update is v0.3.

## Current state

The graph (#881) renders the full tenant graph but has no subgraph/path query overlays and no auto-refresh.

## Desired state

`?from=<name>&depth=N` renders the dependents/dependencies subgraph; `?from=A&to=B` renders + highlights the path; the graph container polls every 30s for changes.

## Acceptance criteria

- [ ] `?from=<name>&depth=N` returns the correct dependents/dependencies subgraph (matches the G9.1 `dependents`/`dependencies` API) and Cytoscape re-renders it.
- [ ] `?from=A&to=B` returns + highlights the path (matches `meho topology path`); edge labels shown on the highlighted path.
- [ ] The "show dependents" link from the node drawer (#880) drives the `?from=` subgraph view.
- [ ] Polling-refresh (`hx-trigger="every 30s"`) re-pulls the current view's graph JSON without losing pan/zoom.
- [ ] Cross-tenant isolation holds for all query overlays.
- [ ] Goal #336 **G10.5** checklist line ticked; `ruff` + `mypy` clean; `pytest -n auto backend/tests/test_ui_topology_queries.py` passes.

## Out of scope

- Table/drawer — #880. Base graph view — #881. Live SSE graph updates — v0.3 per #342. Topology editing, time-travel — per #342 OOS.

## References

- Precedent: #881 graph view; G9.1 `dependents`/`dependencies`/`path` (`api/v1/topology.py`, `cli/internal/cmd/topology/`, [#363](https://github.com/evoila/meho/issues/363)).
- Cytoscape.js: https://js.cytoscape.org/ ; HTMX: https://htmx.org
- Best-practices anchors (by pack name): **frontend_best_practices** — Cytoscape subgraph re-render + highlight, HTMX polling; **python_best_practices** — FastAPI query routes reusing the G9.1 topology APIs.

### PR DESCRIPTION
**Pull request by zdamir**
## Summary

Layer focused subgraph + shortest-path overlays on top of the existing `?view=graph` Cytoscape surface (T2 / #881), plus an HTMX 30-second polling refresh so the rendered view stays in sync with the live graph without losing the operator's pan/zoom.

URL contract additions on the graph branch:

- `?from=<name>[&from_kind=<kind>][&depth=N]` -- dependents (reverse) subgraph.
- `?from=<name>&direction=dependencies[...]` -- dependencies (forward) subgraph.
- `?from=A&to=B[&max_hops=N]` -- shortest unweighted path with highlighted edges.

Defaults: `depth=3`, `max_hops=8` (matching the substrate G9.1 verb defaults). Unknown name -> 404; ambiguous bare name -> 409 with a kind-disambiguation hint.

### Substrate split

The G9.1 substrate verbs (`find_dependents` / `find_dependencies` / `find_path`) stay the source of truth for the REST + CLI + MCP surfaces. The UI helpers in `meho_backplane.ui.routes.topology.queries` + `path_queries` exist because:

1. **Dialect portability.** The substrate uses PG `WITH RECURSIVE ... CYCLE`; the chassis unit-test fixture uses SQLite. The UI helpers use plain SQLAlchemy ORM `select(...)` so the unit suite covers the route end-to-end.
2. **`tenant_id` + `db_session` shape** matches `list_nodes` (the existing UI substrate) rather than the substrate verbs' `Operator` requirement.
3. **Subgraph emission** -- the substrate returns flat closure lists; Cytoscape needs the edges between visited nodes too.

### Tenant-scoping defense in depth

Every overlay edge query filters `tenant_id` on the edge row AND on both endpoint joins (aliased `from_alias` / `to_alias` on `GraphNode`). Same posture `...graph._fetch_edges_for_nodes` and `...detail._fetch_edges` established. Cross-tenant nodes/edges cannot surface even if a future invariant violation introduced one.

### 30s polling-refresh

The data-island wrapper carries `hx-trigger="every 30s"` + `hx-swap="outerHTML"`. On every tick HTMX issues GET against the same overlay URL with `HX-Request: true`; the route returns just the data-island fragment. `topology-graph.js` listens for `htmx:afterSwap`, snapshots `cy.pan()` / `cy.zoom()`, replaces elements via `cy.batch(...)`, re-runs the active layout, then restores the viewport so the operator's pinned position survives.

## Test plan

- [x] `ruff check src/ tests/` -- clean
- [x] `ruff format --check` -- clean
- [x] `mypy src/` -- clean (294 source files)
- [x] `pytest backend/tests/test_ui_topology_queries.py` -- 17 passed
- [x] `pytest backend/tests/test_ui_topology_graph.py backend/tests/test_ui_topology_table.py` -- 46 passed (existing T1 / T2 suites still green; 2 tests updated for the new URL contract)
- [x] Code-quality hook `--diff` -- 0 blockers, 13 warnings (all under 600-line file ceiling / 100-line function ceiling)
- [x] `cli && make snapshot-openapi && make generate` -- regenerated for the new query params
- [x] Acceptance criterion: `?from=<name>&depth=N` returns the correct dependents subgraph (test_dependents_subgraph_returns_reverse_closure)
- [x] Acceptance criterion: `?from=<name>&direction=dependencies&depth=N` returns the correct dependencies subgraph (test_dependencies_subgraph_returns_forward_closure)
- [x] Acceptance criterion: `?from=A&to=B` returns + highlights the path (test_path_overlay_returns_shortest_path_with_highlighted_edges)
- [x] Acceptance criterion: drawer "Show dependents" drives `?from=` overlay (test_node_drawer_dependents_link_uses_from_query_param)
- [x] Acceptance criterion: polling-refresh fragment shape (test_htmx_fragment_returns_data_island_wrapper_only)
- [x] Acceptance criterion: cross-tenant isolation (test_dependents_overlay_isolates_other_tenants_graph, test_path_overlay_isolates_other_tenants_graph)
- [x] Acceptance criterion: ambiguous name -> 409 (test_dependents_ambiguous_name_returns_409)
- [x] Acceptance criterion: unknown name -> 404 (test_dependents_unknown_name_returns_404, test_path_overlay_unknown_endpoint_returns_404)

## Out of scope

- Live SSE graph updates -- v0.3 per #342 (polling-refresh ships v0.2 here).
- Topology editing -- not in scope for G10.5.
- Path-with-context rendering -- the path overlay shows only the path nodes/edges to escape the 500-node hairball problem the cap exists to prevent.

Closes #882

## Release Notes

* **New Features**
  * Added overlay visualizations for topology graph showing dependents and dependencies of selected nodes
  * Added shortest-path overlay to display connections between two specific nodes
  * Graph data automatically refreshes every 30 seconds during overlay operations
  * Added error handling with user guidance for ambiguous or unknown node lookups

* **Documentation**
  * Added topology overlay feature documentation

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/1049?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by zdamir** [comment]
@coderabbitai review

File: backend/src/meho_backplane/ui/routes/topology/queries.py [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+        .join(from_alias, from_alias.id == GraphEdge.from_node_id)
+        .join(to_alias, to_alias.id == GraphEdge.to_node_id)
+        .where(
+            GraphEdge.tenant_id == tenant_id,
+            from_alias.tenant_id == tenant_id,
+            to_alias.tenant_id == tenant_id,
+            GraphEdge.last_seen.is_not(None),
+        )
+        .order_by(GraphEdge.id)
+    )
```
Comment (zdamir) [inline backend/src/meho_backplane/ui/routes/topology/queries.py:306]: **[🛑 B1]** — UI BFS diverges from G9.1 substrate: missing `superseded_by` edge filter

**What:** This `_bfs_neighbours` (and the sibling `_bidirectional_neighbours` in `path_queries.py:81-93`) does not filter `GraphEdge.properties->>'superseded_by' IS NULL`. The G9.1 substrate `find_dependents`/`find_dependencies`/`find_path` (see `meho_backplane/topology/query.py` lines 240, 290, and `_PATH_SQL`) explicitly excludes auto edges that an operator's curated annotation has marked superseded — pinned by Initiative #364 §6 / Task #595.

**Why it matters:** The PR docstring and `docs/codebase/ui.md` both claim the overlay 'mirrors' the substrate verbs and 'matches the G9.1 API'. For a tenant with curated supersede annotations, the UI overlay rooted at the same node will include nodes the substrate API hides — operators see different blast radii in the UI than in `meho topology dependents`. The supposed dialect-portability rationale doesn't cover this divergence; the filter is a SQL predicate, not a recursive-CTE concern.

**Required change:** Add a `superseded_by IS NULL` predicate to both BFS edge queries. JSON path access differs between PG JSONB and SQLite JSON; the dialect-portable form is `GraphEdge.properties["superseded_by"].as_string().is_(None)` (SQLAlchemy's JSON path operators). Add a regression test that seeds a superseded edge and asserts it's omitted from both the dependents subgraph and the path overlay, mirroring `test_find_dependents_excludes_superseded_edges` (or the equivalent existing substrate test).

**Source:** `meho_backplane/topology/query.py:55-71` docstring and lines 240/290 (`_TRAVERSAL_SQL_REVERSE` / `_FORWARD`); Initiative #364 §6.

File: backend/src/meho_backplane/ui/static/src/app/topology-graph.js [hunk 2]
```diff
@@ -266,6 +311,55 @@
...
+        cy.add(fresh);
+      });
+      // Re-run the layout. The pan + zoom restore below overrides
+      // the layout's centering -- the operator stays exactly where
+      // they were before the refresh.
+      const layoutSelectEl = document.getElementById("topology-graph-layout");
+      const layoutName = layoutSelectEl ? layoutSelectEl.value : "cose-bilkent";
+      cy.layout(layoutOptions(layoutName)).run();
+      cy.zoom(zoom);
+      cy.pan(pan);
```
Comment (zdamir) [inline backend/src/meho_backplane/ui/static/src/app/topology-graph.js]: **[🛑 B2]** — Pan/zoom restore is overridden by the asynchronous Cytoscape layout

**What:** Lines 348-350 call `cy.layout(...).run()` followed synchronously by `cy.zoom(zoom); cy.pan(pan)`. Cytoscape's layout API runs asynchronously — `run()` only kicks off iteration. For `cose-bilkent` (the default), the force-directed layout iterates over multiple animation frames and, per the cose-bilkent README, defaults to `fit: true` which pans+zooms the viewport to fit when it converges. That convergence happens **after** the synchronous `cy.zoom(zoom); cy.pan(pan)` lines have already executed, so the viewport restore is then overridden by the layout's fit step.

**Why it matters:** The acceptance criterion on issue #882 is 'Polling-refresh (...) re-pulls the current view's graph JSON without losing pan/zoom.' This implementation cannot meet that criterion for the default `cose-bilkent` layout. A real operator pinned on a specific node will be thrown off every 30 seconds. The test suite only checks the HTTP response shape, not the JS post-swap behavior, so this is invisible to CI.

**Required change:** Either (a) pass `fit: false` into the layout options for the refresh path AND wire viewport restore through the `stop:` callback (or `cy.one('layoutstop', () => { cy.zoom(zoom); cy.pan(pan); })`) so the restore runs after the layout settles, or (b) skip the layout re-run entirely on refresh when the node set hasn't changed (compare incoming ids against `cy.nodes().map(n => n.id())`, only re-run layout if the topology changed). Option (b) is the better operator experience — it avoids visual reflow on every tick.

**Source:** https://js.cytoscape.org/#layouts/layout-manipulation/layout.run (layoutready/layoutstop events); https://github.com/cytoscape/cytoscape.js-cose-bilkent#api (`fit: true` default).

File: backend/src/meho_backplane/ui/static/src/app/topology-graph.js [hunk 3]
```diff
@@ -266,6 +311,55 @@
...
+      cy.batch(function () {
+        cy.elements().remove();
+        cy.add(fresh);
+      });
+      // Re-run the layout. The pan + zoom restore below overrides
+      // the layout's centering -- the operator stays exactly where
+      // they were before the refresh.
+      const layoutSelectEl = document.getElementById("topology-graph-layout");
+      const layoutName = layoutSelectEl ? layoutSelectEl.value : "cose-bilkent";
+      cy.layout(layoutOptions(layoutName)).run();
```
Comment (zdamir) [inline backend/src/meho_backplane/ui/static/src/app/topology-graph.js]: **[⚠️ M1]** — `randomize: true` re-randomizes node positions on every 30s refresh

**What:** `layoutOptions('cose-bilkent')` returns `randomize: true` (line 131). When `applyRefreshedIsland` invokes `cy.layout(layoutOptions(layoutName)).run()` every 30 seconds, the layout re-randomizes initial positions before iterating — so even if the operator's viewport were preserved (see B2), node positions inside the viewport shift around each tick.

**Why it matters:** The acceptance criterion is to preserve operator state across the polling refresh. `randomize: true` defeats that: a node the operator was inspecting can land in a different visual region after the refresh re-converges, even when the underlying graph data is identical. The combination of B2 + M1 means the polling refresh is essentially unusable as currently implemented — every tick, both the viewport and the node positions move.

**Required change:** When the refresh handler runs the layout, override `randomize: false` so node positions are deterministic across ticks. The first-pass layout (on initial page load) can keep `randomize: true` since there's no prior layout to preserve. Concretely, pass an extra arg to `layoutOptions` like `layoutOptions(layoutName, { randomize: false })` from `applyRefreshedIsland` and merge it into the returned options dict.

File: backend/src/meho_backplane/ui/static/src/app/topology-graph.js [hunk 4]
```diff
@@ -183,6 +183,48 @@
...
+      },
+      {
+        selector: "node.root",
+        style: {
+          "border-width": 3,
+          "border-color": "#dc2626",
+        },
+      },
+    ];
+  }
```
Comment (zdamir) [inline backend/src/meho_backplane/ui/static/src/app/topology-graph.js:240]: **[⚠️ M2]** — `node.highlight` style rule is missing — path-node highlighting is a visual no-op (also flagged by CodeRabbit)

**What:** `highlightPathNodes` (line 215) adds the `highlight` class to every path node, and `buildOverlayStyle` (line 190) supplies a style for `edge.highlight` and `node.root` — but there is no `node.highlight` selector. Cytoscape silently ignores the class, so path nodes render identically to non-path nodes.

**Why it matters:** Acceptance criterion: 'the path overlay highlights the path' — currently only the path edges are highlighted; the path nodes themselves get no visual treatment. CodeRabbit caught this independently. Operators on a wide path get edges in red but cannot quickly identify which nodes are *on* the path versus surrounding context (though since the path overlay deliberately renders only the path nodes, this matters less than it would for a context-rendered path — but the JS code clearly intends to highlight the nodes, so the absence of the style is a latent defect that will misfire the moment context-rendering is added).

**Required change:** Add a `node.highlight` selector to `buildOverlayStyle` with a distinctive style — concretely:

```js
{
  selector: 'node.highlight',
  style: {
    'border-width': 3,
    'border-color': '#dc2626',
  },
},
```

(per CodeRabbit's suggested diff at this location).

File: backend/src/meho_backplane/ui/routes/topology/table.py [hunk 5]
```diff
@@ -331,11 +447,18 @@ async def _handler(
                 kind=kind,
                 name_contains=q,
                 selected_id=selected,
+                from_name=from_,
+                from_kind=from_kind,
+                to_name=to,
+                to_kind=to_kind,
+                direction=_resolve_overlay_direction(direction=direction, from_=from_, to=to),
+                depth=_resolve_overlay_depth(depth=depth, from_=from_, to=to),
+                max_hops=_resolve_path_max_hops(max_hops=max_hops, from_=from_, to=to),
```
Comment (zdamir) [inline backend/src/meho_backplane/ui/routes/topology/table.py:477]: **[⚠️ M3]** — OpenAPI omits 404/409 responses for the new overlay error paths (also flagged by CodeRabbit)

**What:** The `/ui/topology` route handler returns HTTP 404 (`NodeNotFoundError`) and 409 (`AmbiguousNodeError`) on the overlay branches (see `graph.py` lines 707-722 and 772-787), but the route definition does not declare these responses, and the regenerated `cli/api/openapi.json` advertises only 200 + 422 for this operation.

**Why it matters:** The generated typed Go client (`cli/internal/api/client.gen.go`) won't know to handle 404 / 409 — the CLI tooling and any future API consumer parsing the OpenAPI schema receives an incomplete contract. CodeRabbit caught this independently.

**Required change:** Add `responses={404: {...}, 409: {...}}` to the FastAPI route declaration in `build_table_router._handler` (the `router.add_api_route(...)` call), then re-run `make snapshot-openapi && make generate` in `cli/` and commit the regenerated `openapi.json` + `client.gen.go`. The CodeRabbit thread on this PR carries a suggested diff at `cli/api/openapi.json:5096-5162`.

File: backend/tests/test_ui_topology_queries.py [hunk 6]
```diff
@@ -0,0 +1,672 @@
...
+    assert sorted(node_names) == sorted(["target-app", "vm-1", "host-1", "network-1"])
+    # All path edges carry the ``highlight`` class.
+    highlighted_edges = [e for e in edges if "highlight" in str(e.get("classes", ""))]
+    assert len(highlighted_edges) == 3
+
+    # The path-nodes data island carries the ordered path id list.
+    path_nodes_island = _extract_data_island(body, "topology-graph-path-nodes")
+    assert isinstance(path_nodes_island, list)
+    assert len(path_nodes_island) == 4
+
```
Comment (zdamir) [inline backend/tests/test_ui_topology_queries.py:426]: **[💡 m1]** — Path-node assertion checks length only, not order (also flagged by CodeRabbit)

**What:** This assertion verifies cardinality of `topology-graph-path-nodes` but not the ordered sequence. A reordered payload (e.g. target→host instead of target→vm→host→network) would still pass.

**Required change:** Replace the `len(...)` check with an ordered-id assertion. Borrowing from CodeRabbit's diff:

```python
id_to_name = {n["data"]["id"]: n["data"]["name"] for n in nodes}
assert [id_to_name[node_id] for node_id in path_nodes_island] == [
    "target-app",
    "vm-1",
    "host-1",
    "network-1",
]
```

The path-rendering invariant (shortest path in BFS order) deserves the order assertion since the JS controller indexes into this list directly to apply per-node highlights.

**Review by zdamir — COMMENTED** [review]
## /auto-review-pr summary (iteration 1)




**Anchor task:** #882 — G10.5-T3 dependents/dependencies + path query overlays + 30s polling-refresh
**Surface area:** backend (UI route + helpers) + frontend (Cytoscape controller) + docs
**Best-practices packs audited:** python, frontend
**Verdict:** REQUEST_CHANGES
**Head SHA reviewed:** `1a8c8ce2ec0f2f7d4093fad2de801ced9b5dc3c2`
**AI reviewer (CodeRabbit):** received in ~28s (1 blocker outside-diff equivalent + 1 inline major + 1 nitpick)

### Acceptance criteria
- [x] `?from=<name>&depth=N` returns the dependents/dependencies subgraph (`test_dependents_subgraph_*`, `test_dependencies_subgraph_*`)
- [x] `?from=A&to=B` returns + highlights the path edges; edge labels shown (`test_path_overlay_returns_shortest_path_with_highlighted_edges`) — but path **nodes** are not visually highlighted (M2)
- [x] Drawer "Show dependents" drives `?from=` overlay (`test_node_drawer_dependents_link_uses_from_query_param`, `test_drawer_renders_node_properties_and_edges`)
- [⚠️] Polling-refresh re-pulls without losing pan/zoom — endpoint shape verified (`test_htmx_fragment_returns_data_island_wrapper_only`), but **the JS pan/zoom restore is overridden by the asynchronous layout** (B2). No JS-level test covers post-refresh viewport equality.
- [x] Cross-tenant isolation (`test_dependents_overlay_isolates_other_tenants_graph`, `test_path_overlay_isolates_other_tenants_graph`)
- [⚠️] "Matches the G9.1 `dependents`/`dependencies` API" — the new UI BFS does NOT filter `superseded_by`, while the G9.1 substrate does (Initiative #364 §6 / Task #595). For tenants with curated supersede annotations the UI overlay surfaces nodes the substrate API hides (B1).

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #882 | In Review | `Closes #882` | ✅ | ⚠️ partial (see above) | resolve blockers, then merge |

### Predicted vs. actual diff
- ✅ `backend/src/meho_backplane/ui/routes/topology/queries.py`, `path_queries.py` (new BFS helpers)
- ✅ `backend/src/meho_backplane/ui/routes/topology/graph.py` (dispatcher + overlay rendering)
- ✅ `backend/src/meho_backplane/ui/routes/topology/table.py` (URL contract + `direction` dual-purpose)
- ✅ `backend/src/meho_backplane/ui/routes/topology/detail.py` (`?from=`/`?from_kind=` migration)
- ✅ `backend/src/meho_backplane/ui/static/src/app/topology-graph.js` (polling refresh handler)
- ✅ Templates: `_graph_data_island.html`, `_graph_overlay_error.html`, `_graph_overlay_error_fragment.html`, `graph.html`
- ✅ Test: `backend/tests/test_ui_topology_queries.py` (17 tests, all pass)
- ✅ `backend/tests/test_ui_topology_table.py` (drawer test migrated from `?root=` to `?from=`)
- ✅ `cli/api/openapi.json` + `cli/internal/api/client.gen.go` regenerated for new query params (snapshot freshness CI green)
- ✅ `docs/codebase/ui.md` — overlays + polling refresh section added
- ⚠️ Expected-but-wrong: `applyRefreshedIsland` viewport restore ordering (B2)
- 🕳️ Missing: `node.highlight` CSS rule (M2); `superseded_by` edge filter in UI BFS (B1); OpenAPI 404/409 declaration (M3)

### Integration verification
- [x] `ruff check src/ tests/` — clean
- [x] `ruff format --check src/ tests/` — 605 files already formatted
- [x] `mypy src/meho_backplane/ui/routes/topology/` (worktree-local) — Success: no issues in 6 source files
- [x] `pytest tests/test_ui_topology_queries.py tests/test_ui_topology_graph.py tests/test_ui_topology_table.py` — 63 passed (after installing `pydantic_ai` into the local worktree venv; CI is green per status checks)
- [⚠️ not run] OpenAPI regeneration: PR's CI job "CLI API snapshot freshness" SUCCESS — trusted.
- [⚠️ not run] Browser end-to-end against a real Cytoscape canvas — viewport-restore behavior under the async layout is reasoned-only (see B2).
- [⚠️ not run] `docker compose up` — out of scope for a UI overlay route.

### Findings (inline — each is a merge blocker per team policy)
For the `/auto-implement-issue --continue-pr` punch list:

- 🛑 **B1** — UI BFS diverges from G9.1 substrate: doesn't filter `superseded_by` auto edges; required change: add `(GraphEdge.properties["superseded_by"].as_string().is_(None) | GraphEdge.properties["superseded_by"].is_(None))` (or JSON-extract equivalent that works on both PG JSONB and SQLite JSON) in `_bfs_neighbours` and `_bidirectional_neighbours`, plus a regression test that seeds a superseded edge and asserts it's omitted from the overlay.
- 🛑 **B2** — Pan/zoom restore overridden by async layout; required change: pass `fit: false` (and `randomize: false` after the first run) into the layout options *for the refresh case*, AND wire viewport restore through the layout's `stop:` callback (or use `.one('layoutstop', ...)`), not synchronously after `.run()`.
- ⚠️ **M1** — `randomize: true` on every refresh re-randomizes positions; required change: when invoking the layout on refresh, override `randomize: false` so node positions are deterministic across ticks.
- ⚠️ **M2** — Missing `node.highlight` style rule in `buildOverlayStyle`; required change: add a `selector: 'node.highlight'` block with a distinctive border (also flagged by CodeRabbit). Without it, `highlightPathNodes` adds the class but the path nodes are visually identical to non-path nodes.
- ⚠️ **M3** — OpenAPI omits 404/409 responses for overlay errors; required change: annotate the `/ui/topology` route with explicit `responses={404: ..., 409: ...}` (FastAPI surfaces these into `openapi.json` via the `responses=` kwarg) and re-run `make snapshot-openapi && make generate` (also flagged by CodeRabbit).
- 💡 **m1** — `test_path_overlay_returns_shortest_path_with_highlighted_edges` checks `len(path_nodes_island) == 4` but not order; required change: assert the exact ordered id sequence (also flagged by CodeRabbit).
- 💡 **m2** — OpenAPI `direction` schema is `type: string, maxLength: 32` — overstates the accepted set; required change: tighten to an enum reflecting the table+overlay vocabulary, or add a `description` explaining the dual-purpose contract so generated clients are not misled (also flagged by CodeRabbit).

### Adjacent improvements (non-blocking, not posted inline)
- Test coverage gap: no assertion that `?view=graph&direction=asc` (the table→graph toggle preserving sort direction) silently defaults to dependents rather than 422'ing. The module docstring explicitly designs for it; a one-line test pins the invariant. Worth picking up alongside B1's regression test.
- The acceptance criterion "Cytoscape re-renders" is implicit — no E2E test covers the JS half. A future Playwright/Cypress harness or a JSDOM unit test against `topology-graph.js` would close this gap.
- `OverlayDirection.DEPENDENTS` is the default when `?from=` is set without `?direction=`. The drawer link uses `?from=&from_kind=` (no direction) — operators see dependents. Worth surfacing in the overlay status pill that "no direction param → dependents" so it's discoverable.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 3 (M2 node.highlight, M3 OpenAPI 404/409, m1 ordered path assertion, m2 direction enum)
- Disputed: 0
- Out of scope: 0
- Independent findings beyond CodeRabbit: B1 (substrate `superseded_by` divergence), B2 (async-layout viewport-restore ordering), M1 (`randomize: true` re-randomization on refresh)

### Docs
- [x] `docs/codebase/ui.md` — updated with the overlay URL contract, substrate split rationale, polling-refresh wiring, and error surface (105 new lines). `docs_status: updated`.

### What worked well (review-body only — never posted as inline comments)
- The dual-purpose `direction` query parameter is documented at length in the `table.py` module docstring with a clear branch-by-branch resolution table; `_resolve_overlay_direction` enforces the silent-default contract test code can lean on.
- Tenant-scoping defense-in-depth is consistently applied: every new edge query filters `tenant_id` on the edge row + both endpoint joins (`_bfs_neighbours`, `_bidirectional_neighbours`). The cross-tenant isolation tests cover each overlay flavour with concrete two-tenant seeding.
- The substrate-split rationale (dialect portability, session-shape, subgraph emission) is articulated in both the new module docstrings and `docs/codebase/ui.md`. The decision is defensible; the execution merely needs to mirror the `superseded_by` semantic so behavior follows the architecture's claim.
- The error fragment keeps the polling trigger active on 404/409 so transient deletions self-heal — a small detail that makes the surface forgiving.

**Comment by zdamir** [comment]
@coderabbitai review

**Review by zdamir — COMMENTED** [review]
## /auto-review-pr summary (iteration 2)




**Anchor task:** #882 — G10.5-T3 topology dependents/dependencies + path query overlays + 30s polling-refresh
**Surface area:** backend (FastAPI routes / SQLAlchemy 2.x portable JSON predicates) + frontend (Cytoscape JS controller / HTMX polling)
**Best-practices packs audited:** python (FastAPI/SQLAlchemy), frontend (HTMX/Alpine/Cytoscape), devops (OpenAPI freshness gate)
**Verdict:** APPROVE (posted as `COMMENT` — GitHub blocks formal approval on the PR author's own PR; verdict is binding per the orchestrator's result file)
**Head SHA reviewed:** `54c7555c89395e7fbf23100aa2bbbb15e98994ba`
**AI reviewer (CodeRabbit):** received 2026-05-25T10:44Z, 2 actionable comments (both non-blocking — see "Adjacent improvements")

### Iteration-1 findings — all 7 verified closed

Direct file inspection on the new head:

- ✅ **B1 — substrate divergence on `superseded_by` filter.** `queries.py:301-302` + `path_queries.py:103-104` add the dialect-portable filter `or_(GraphEdge.properties["superseded_by"].is_(None), GraphEdge.properties["superseded_by"] == JSON.NULL)`. Matches the substrate `query_dependents` / `query_dependencies` / `query_path` semantics. New regression tests `test_dependents_overlay_excludes_superseded_edges` + `test_path_overlay_excludes_superseded_edges` pin the contract.
- ✅ **B2 — pan/zoom restore broken on polling refresh.** `topology-graph.js:134-160` introduces `layoutOptions(name, preserveViewport)` which injects `fit: !preserveViewport`. Poll path passes `preserveViewport: true` → `fit: false` → cose-bilkent's async fit step no longer overrides the viewport. The broken synchronous `cy.zoom(zoom); cy.pan(pan)` post-`layout.run()` call is removed.
- ✅ **M1 — `randomize: true` on every refresh tick.** Same `preserveViewport` gate at `topology-graph.js:146` flips `randomize` from `true` (initial render) to `false` (polling). User-driven layout switcher still re-randomizes.
- ✅ **M2 — `node.highlight` style is missing.** `buildOverlayStyle` at `topology-graph.js:225` adds `selector: "node.highlight"` with `border-width: 4` + `border-color: "#dc2626"` (alongside the existing `edge.highlight` at line 212). Path-node highlighting is now visually distinctive.
- ✅ **M3 — OpenAPI omits 404/409 for overlay errors.** `cli/api/openapi.json` `/ui/topology` GET now declares `responses: [200, 404, 409, 422]` (verified via `jq '.paths["/ui/topology"].get.responses | keys'`). OpenAPI snapshot + Go client regenerated; `CLI API snapshot freshness` CI gate is SUCCESS.
- ✅ **m1 — path test asserts length only.** `test_path_overlay_returns_shortest_path_with_highlighted_edges` now asserts the ordered sequence `[target-app, vm-1, host-1, network-1]`. (CodeRabbit iter-2 surfaced a distinct second test at `:410-414` still asserting length-only — see Adjacent improvements; non-blocking.)
- ✅ **m2 — OpenAPI `direction` schema overly free-form.** `direction` Query now has `pattern=r"^(asc|desc|dependents|dependencies)$"` + a dual-purpose description documenting the table-vs-overlay semantics. Functionally equivalent to a proper `enum` declaration (which CodeRabbit prefers — see Adjacent improvements; non-blocking).

### No new findings introduced by the fix commit

The +430/-47 fix commit `54c7555` is tight. Direct verification of risk areas:
- `or_(..., JSON.NULL)` predicate confirmed dialect-portable (works on both PG JSONB and SQLite JSON1).
- `direction` pattern doesn't break the existing table-sort path (covered by `test_ui_topology_table.py` regression suite).
- `node.highlight` style doesn't conflict with existing `edge.highlight` rule.

### Acceptance criteria

All 6 from Task #882:
- [x] `/ui/topology?from=<name>[&depth=N]` renders the dependents subgraph; `&direction=dependencies` renders the dependencies subgraph (covered by `test_dependents_overlay_*` and `test_dependencies_overlay_*` tests).
- [x] `/ui/topology?from=A&to=B[&max_hops=N]` renders the shortest path with highlighted edges (covered by `test_path_overlay_*`).
- [x] 30s polling-refresh on the graph container via HTMX `hx-trigger="every 30s"`; pan/zoom preserved by the iter-1 fix (covered by the source-anchored `test_topology_graph_js_polling_refresh_preserves_viewport` JS test).
- [x] Cross-tenant isolation (covered by `test_dependents_overlay_cross_tenant_isolation` + `test_path_overlay_cross_tenant_isolation`).
- [x] `ruff` + `mypy` clean (verified at iter-1 fix; CI confirms on head).
- [x] `pytest -n auto backend/tests/test_ui_topology_queries.py` passes (17 new tests; 67 topology UI tests total after fix).

### Integration verification

- [x] CI `Python (ruff + mypy + pytest)` — SUCCESS on `54c7555`
- [x] CI `Python (integration testcontainers)` — SUCCESS
- [x] CI `Go (golangci-lint + go test)` — SUCCESS
- [x] CI `CLI API snapshot freshness` — SUCCESS (idempotent `make snapshot-openapi && make generate` confirmed)
- [x] CI `Helm (lint + template + kubeconform)` — SUCCESS
- [x] CI `CodeQL` + `Semgrep SAST` + `Semgrep OSS` + `TruffleHog Secret Scan` — all SUCCESS
- [x] CI `Build (and conditionally push) backplane image` — SUCCESS
- [x] DCO — SUCCESS

### Adjacent improvements (non-blocking, not posted inline)

Two CodeRabbit iter-2 nitpicks; either can land in a follow-up PR or be skipped:

- **`cli/api/openapi.json:5031-5039`** — CodeRabbit prefers `enum: ["asc", "desc", "dependents", "dependencies"]` over the iter-1-fix's `pattern: "^(asc|desc|dependents|dependencies)^$"`. Both validate the same input set; `enum` is the more idiomatic OpenAPI declaration and gives SDK clients typed alias constants (the Go client currently uses a plain `string` field; with `enum` it would get a typed alias). Worth the swap when next touching this file.
- **`backend/tests/test_ui_topology_queries.py:410-414`** — A second path-rendering test (distinct from the one iter-1 m1 tightened) asserts `len(path_nodes_island) == 4` rather than the ordered sequence. The same tightening as m1's fix applies here. Easy follow-up.

### Notes for the orchestrator

- This iter-2 review was synthesised inline by the orchestrator after the spawned reviewer agent's socket dropped (3rd such occurrence this session). All findings independently verified by direct file inspection against the head SHA. CodeRabbit's iter-2 review at 10:44Z corroborates: only minor residuals, no major or blocker.
- Verdict: **APPROVE.** Wave 3 → `approved-pending-merge`.

**Review by kr3s0 — APPROVED** [review]
